### ISSUE TITLE
EpoxyAdapter holds reference to view

### ISSUE THREAD

**Issue by audkar** [airbnb/epoxy#97]
Currently EpoxyAdapter holds strong reference to views. Does this means that Adapter can't be field in Fragment and should be null'ed on `onViewDestroyed`? Ref is hold to Views by [boundViewHolders](https://github.com/airbnb/epoxy/blob/master/epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyAdapter.java#L39)

e.g.
```
Fragment1 {
  val adapter: EpoxyAdapter 

  @override fun onCreateView(...) {
    <..recycler setup..>
  }
}
```
And I go to next Fragment, then View of Fragment1 is leaked until whole fragment is Destroyed.

**Comment by elihart** [comment]
Thanks for pointing this out, I believe you're right. We tend to not use fragment back stacks much so it hasn't really been a problem for us, but this memory leak isn't ideal.

It's fine to have the adapter as a field. The recycler view just needs to recycle the views. There are two ways to force this in `onDestroyView`:

1) Detach the adapter from the recyclerview: `recyclerView.setAdapter(null)`
2) Clear all the models in the adapter and notify the change.

I prefer method one. Unfortunately `onDestroyView` is called before the exiting fragment finishes its transition, so the recyclerview will change to be empty while the fragment is still visible and transitioning out. I don't have a great solution to that, I have handled it by posting a delayed runnable to do `recyclerView.setAdapter(null)` so it isn't cleared until after the fragment transition finishes. It's an unfortunate hack, and I would love to hear better ideas.

I would prefer to avoid the complexity of making the references weak, but if this is a major issue we can explore that solution.

**Comment by audkar** [comment]
> this memory leak isn't ideal

Very gently said :)

> I would prefer to avoid the complexity of making the references weak, but if this is a major issue we can explore that solution.

I can think two solutions. One easy&hacky and second - requires more work&breaks lib behavior :)

1st solution. Just add some flag which turns on/off this state saving feature. And document it heavily, that client has to manually take additional steps to clear adapter when using state saving.

2nd solution. IMHO it is not adapters responsibility to manage views states. Probably it belongs to `LayoutManager`. This is indicated by the fact that it has methods like [OnSaveInstanceState](https://developer.android.com/reference/android/support/v7/widget/RecyclerView.LayoutManager.html#onSaveInstanceState())

Would you consider possibility that this child state saving behavior can be moved  to other classes? This would require some refactoring and manual upgrade steps from lib users.

**Comment by elihart** [comment]
I realized there is a much better way to force the views to be recycled instead of having to clear the adapter. On Linear/GridLayoutManager you just call `setRecycleChildrenOnDetach(true)`. Then when the RecyclerView is detached from the window it will recycle all of its child views, calling through to EpoxyAdapter to unbind all the views.

This seems like a nice way to handle it, at least for now. You're right that it doesn't seem right for the adapter to manage save state. I could spend more time looking into how to decouple it, but afaik state needs to be restored in `bindViewHolder` and only the adapter has good access to that. I would prefer not to have a custom LayoutManager since that makes set up more difficult and also makes people unable to use their own custom layout managers if they need to.

I could change saved state to be turned off by default, and add a setting for it. Before making that kind of breaking change I'll need to think a bit more on it.

The EpoxyAdapter could also be made to automatically call `setRecycleChildrenOnDetach(true)` if a LinearLayoutManager is used. Something like
```
@Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        if (recyclerView.getLayoutManager() instanceof LinearLayoutManager) {
            ((LinearLayoutManager) recyclerView.getLayoutManager()).setRecycleChildrenOnDetach(true);
        }
}
```

That seems a bit heavy handed and not a responsibility of the adapter. We use a base adapter that extends `EpoxyAdapter` to add some custom functionality, I'm going to make this change for our team's usage at least, and that could be an option for other people.

Epoxy does set up an expectation of symmetry in the bind/unbind calls though. Some people may always want their views unbound and properly recycled in order to clean up resources. Having Epoxy auto enable `setRecycleChildrenOnDetach` makes sense in the context of the library and is probably what almost everybody would want and expect. We could have a setting to disable that behavior if needed. 

Thoughts?

**Comment by audkar** [comment]
> I realized there is a much better way to force the views to be recycled instead of having to clear the adapter. On Linear/GridLayoutManager you just call `setRecycleChildrenOnDetach(true)`.

Indeed this is probably best solution in terms of cost to workaround this issue. 

---
However seems that adapter still leaks layouts. I created [example](https://github.com/audkar/EpoxyLeakExample) project with leakcanary setup to demonstrate situation. 
But I didn't had time to dig deeper why adapter still has registered data change listener.

**Comment by elihart** [comment]
Thank you for creating the example project. At first glance I'm not sure why it would be leaking. The only thing I would change is to remove the listener from the view in `unbind` of the model. I don't think not doing that would cause it to leak though.

I'll run the project later and take a look at the leak canary stack trace in depth, unless you're able to post it here first.

Thanks!

**Comment by elihart** [comment]
@audkar So I dug into this, very interesting! This isn't just an Epoxy issue, it will happen with a normal recyclerview adapter - I tried that here https://github.com/elihart/EpoxyLeakExample/blob/master/app/src/main/java/com/example/audrius/testepoxy/FragmentOne.java

The issue is that when an adapter is registered with a recyclerview the recyclerview adds a listener to that adapter. So the adapter will continue to reference the recyclerview via its listener list until the adapter is removed from the recyclerview.

This means that any time a recyclerview is used it would leak if the adapter is a field and it is not removed from the fragment.

I was surprised by this, but I guess it makes sense. It means that we have several leaks in our own app.

The solution is to remove the adapter from the recyclerview when the fragment view is destroyed. Unfortunately that has the problem of clearing the view during fragment transaction like I mentioned in a previous post. I will look into ways to avoid that.

Thanks again for pointing this out! I will update the readme soon for instructions on how to avoid these leaks.

**Comment by elihart** [comment]
@audkar Do you think https://github.com/airbnb/epoxy/pull/106 sufficiently explains the situation?

**Comment by tpesce** [comment]
@elihart I don't use epoxy, but I ran into the same problem with the RecyclerView memory leak and some searching lead me here. I had also tried clearing the adapter, and like you, I found that it interrupted my custom fragment transition animations. For me, even introducing a delay wasn't working.

After more digging I finally found a workaround that solves all of my problems and I wanted to share. Take a look here: http://stackoverflow.com/a/40302176/6872885.
